File: apps/spell-circle-canvas/src/common/compose/reference/examples/filter_verb.py
"""filter — the node's own rendered layer post-processed: the whole
subtree, its children included, goes through the filter.

The Python twin of filter_verb.cpp. A reference example: it is rendered
with ``sigil render`` and belongs to no sketch registry.
"""

from sigil import compose, material
from sigil.sketch import SketchContext, sketch

GROUND = "#0f1318"
PLATE = "#1d2730"
INK = "#9fe3d4"
ASH = "#8ea0ad"


def cell(caption: str, plate: compose.Element) -> compose.Element:
    """The same subtree — a plate, a word and a rule — under each
    filter."""
    return (
        compose.box(
            plate.width(compose.pct(100))
            .height(120)
            .borderRadius(10)
            .fill(PLATE)
            .column()
            .gap(8)
            .justifyContent("center")
            .alignItems("center")
            .children(
                compose.text("SIGNAL", size=26, color=INK).fontTrack(3),
                compose.box().width(58).height(4).fill(INK),
            ),
            compose.text(caption, size=12, color=ASH),
        )
        .column()
        .gap(10)
        .flexBasis(0)
        .flexGrow(1)
        .alignItems("center")
    )


@sketch(size=(640, 250), background=GROUND, capture_at=0)
class FilterVerb:
    def setup(self, ctx: SketchContext) -> None:
        ctx.render(self.describe())

    def describe(self) -> compose.Element:
        return (
            compose.box(
                cell("no filter", compose.box()),
                cell(
                    "filter(blur)", compose.box().filter(material.Effect.blur(3))
                ),
                cell(
                    "filter(glow)",
                    compose.box().filter(material.Effect.glow("#3fd6b0", 9)),
                ),
            )
            .row()
            .gap(16)
            .padding(24)
        )

File: apps/spell-circle-canvas/src/common/compose/reference/examples/filter_verb.cpp
/** @file
 * filter — the node's own rendered layer post-processed: the whole
 * subtree, its children included, goes through the filter.
 *
 * A reference example. It stands outside the sketch registry, so it is
 * photographed with `--frame` and never enters the plate sweep.
 */

#include <sigilcompose/core/Core.h>
#include <sigilmaterial/color/Color.h>
#include <sigilmaterial/skia/Effect.h>
#include <sigilsketch/canvas/Sketch.h>

namespace material = sigil::material;
namespace sketch = sigil::sketch;
namespace skia = sigil::material::skia;

using namespace sigil::compose;

namespace {

constexpr SkSize kCanvas = {640, 250};
constexpr material::Color kGround = hexColor(0x0f1318);
constexpr material::Color kPlate = hexColor(0x1d2730);
constexpr material::Color kInk = hexColor(0x9fe3d4);
constexpr material::Color kAsh = hexColor(0x8ea0ad);

/** The same subtree — a plate, a word and a disc — under each filter. */
Element cell(const char* caption, Element plate) {
  return box()
      .column()
      .gap(10)
      .flexBasis(0)
      .flexGrow(1)
      .alignItems(Align::Center)
      .children({plate.width(pct(100))
                     .height(120)
                     .borderRadius({10})
                     .fill(kPlate)
                     .column()
                     .gap(8)
                     .justifyContent(Justify::Center)
                     .alignItems(Align::Center)
                     .children({text("SIGNAL").font(
                                    {.size = 26,
                                     .color = material::skia::toSkColor(kInk),
                                     .track = 3}),
                                box().width(58).height(4).fill(kInk)}),
                 text(caption).font(
                     {.size = 12, .color = material::skia::toSkColor(kAsh)})});
}

}  // namespace

struct FilterVerb {
  void setup(sketch::SketchContext& ctx) {
    ctx.canvas({.size = kCanvas, .background = kGround, .captureSeconds = 0});
    ctx.composer.render(describe());
  }

  Element describe() const {
    return box().row().gap(16).padding(24).children({
        cell("no filter", box()),
        cell("filter(blur)", box().filter(skia::Effect::blur(3))),
        cell("filter(glow)",
             box().filter(skia::Effect::glow(hexColor(0x3fd6b0), 9))),
    });
  }
};

SIGIL_SKETCH(FilterVerb, "Reference · Compose",
             "one subtree drawn plain, blurred, and glowing")
